In [1]:
#Importing required packages
import pandas as pd
from sklearn import __version__ as sklearn_version
from sklearn.preprocessing import scale
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler

In [2]:
#Load the data
df_applicant= pd.read_csv(r'C:\Users\gibib\Desktop\Data Scientist\Springboard-Project-Capstone-2\Credit Card Data/Applicant_data_explored.csv')

In [3]:
df_applicant.head()

,ID,CODE_GENDER,FLAG_OWN_CAR,FLAG_OWN_REALTY,CNT_CHILDREN,AMT_INCOME_TOTAL,NAME_INCOME_TYPE,NAME_EDUCATION_TYPE,NAME_FAMILY_STATUS,NAME_HOUSING_TYPE,DAYS_BIRTH,DAYS_EMPLOYED,FLAG_WORK_PHONE,FLAG_PHONE,FLAG_EMAIL,CNT_FAM_MEMBERS,STATUS
0,5008804,M,1,1,0,427500.0,Working,Higher education,Civil marriage,Rented apartment,-12005,-4542.0,1,0,0,2,0
1,5008805,M,1,1,0,427500.0,Working,Higher education,Civil marriage,Rented apartment,-12005,-4542.0,1,0,0,2,0
2,5008806,M,1,1,0,112500.0,Working,Secondary / secondary special,Married,House / apartment,-21474,-1134.0,0,0,0,2,1
3,5008808,F,0,1,0,270000.0,Commercial associate,Secondary / secondary special,Single / not married,House / apartment,-19110,-3051.0,0,1,1,1,1
4,5008809,F,0,1,0,270000.0,Commercial associate,Secondary / secondary special,Single / not married,House / apartment,-19110,-3051.0,0,1,1,1,1


In [4]:
df_applicant.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 36457 entries, 0 to 36456
Data columns (total 17 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   ID                   36457 non-null  int64  
 1   CODE_GENDER          36457 non-null  object 
 2   FLAG_OWN_CAR         36457 non-null  int64  
 3   FLAG_OWN_REALTY      36457 non-null  int64  
 4   CNT_CHILDREN         36457 non-null  int64  
 5   AMT_INCOME_TOTAL     36457 non-null  float64
 6   NAME_INCOME_TYPE     36457 non-null  object 
 7   NAME_EDUCATION_TYPE  36457 non-null  object 
 8   NAME_FAMILY_STATUS   36457 non-null  object 
 9   NAME_HOUSING_TYPE    36457 non-null  object 
 10  DAYS_BIRTH           36457 non-null  int64  
 11  DAYS_EMPLOYED        36457 non-null  float64
 12  FLAG_WORK_PHONE      36457 non-null  int64  
 13  FLAG_PHONE           36457 non-null  int64  
 14  FLAG_EMAIL           36457 non-null  int64  
 15  CNT_FAM_MEMBERS      36457 non-null 

In [5]:
#To find out the category columns
df_applicant.select_dtypes(include=['object'])

,CODE_GENDER,NAME_INCOME_TYPE,NAME_EDUCATION_TYPE,NAME_FAMILY_STATUS,NAME_HOUSING_TYPE
0,M,Working,Higher education,Civil marriage,Rented apartment
1,M,Working,Higher education,Civil marriage,Rented apartment
2,M,Working,Secondary / secondary special,Married,House / apartment
3,F,Commercial associate,Secondary / secondary special,Single / not married,House / apartment
4,F,Commercial associate,Secondary / secondary special,Single / not married,House / apartment
...,...,...,...,...,...
36452,M,Working,Secondary / secondary special,Married,House / apartment
36453,F,Commercial associate,Higher education,Married,House / apartment
36454,F,Pensioner,Higher education,Married,House / apartment
36455,F,Working,Secondary / secondary special,Married,House / apartment


In [6]:
#To find out the count of each category
counts_CODE_GENDER =df_applicant['CODE_GENDER'].value_counts()
print(counts_CODE_GENDER)

CODE_GENDER
F    24430
M    12027
Name: count, dtype: int64


In [7]:
#To find out the count of each category
counts_NAME_INCOME_TYPE =df_applicant['NAME_INCOME_TYPE'].value_counts()
print(counts_NAME_INCOME_TYPE)

NAME_INCOME_TYPE
Working                 18819
Commercial associate     8490
Pensioner                6152
State servant            2985
Student                    11
Name: count, dtype: int64


In [8]:
#To find out the count of each category
counts_NAME_EDUCATION_TYPE=df_applicant['NAME_EDUCATION_TYPE'].value_counts()
print(counts_NAME_EDUCATION_TYPE)

NAME_EDUCATION_TYPE
Secondary / secondary special    24777
Higher education                  9864
Incomplete higher                 1410
Lower secondary                    374
Academic degree                     32
Name: count, dtype: int64


In [9]:
#To find out the count of each category
counts_NAME_FAMILY_STATUS=df_applicant['NAME_FAMILY_STATUS'].value_counts()
print(counts_NAME_FAMILY_STATUS)

NAME_FAMILY_STATUS
Married                 25048
Single / not married     4829
Civil marriage           2945
Separated                2103
Widow                    1532
Name: count, dtype: int64


In [10]:
#To find out the count of each category
counts_NAME_HOUSING_TYPE=df_applicant['NAME_HOUSING_TYPE'].value_counts()
print(counts_NAME_HOUSING_TYPE)

NAME_HOUSING_TYPE
House / apartment      32548
With parents            1776
Municipal apartment     1128
Rented apartment         575
Office apartment         262
Co-op apartment          168
Name: count, dtype: int64


For the columns 'NAME_EDUCATION_TYPE' and 'NAME_HOUSING_TYPE', few categories are having very less no.s compared to others. So instead of specifying all categories, we can specify these categories together as 'Other'. So that we can limit the no. of columns, when creating dummies.

In [11]:
#To change less occuring categories as 'other'
mask_EDU=df_applicant['NAME_EDUCATION_TYPE'].isin(counts_NAME_EDUCATION_TYPE[counts_NAME_EDUCATION_TYPE<1000].index)
df_applicant['NAME_EDUCATION_TYPE'][mask_EDU]='Other'
print(df_applicant['NAME_EDUCATION_TYPE'].value_counts())

NAME_EDUCATION_TYPE
Secondary / secondary special    24777
Higher education                  9864
Incomplete higher                 1410
Other                              406
Name: count, dtype: int64


C:\Users\gibib\AppData\Local\Temp\ipykernel_14552\3122443786.py:3: FutureWarning: ChainedAssignmentError: behaviour will change in pandas 3.0!
You are setting values through chained assignment. Currently this works in certain cases, but when using Copy-on-Write (which will become the default behaviour in pandas 3.0) this will never work to update the original DataFrame or Series, because the intermediate object on which we are setting values will behave as a copy.
A typical example is when you are setting values in a column of a DataFrame, like:

df["col"][row_indexer] = value

Use `df.loc[row_indexer, "col"] = values` instead, to perform the assignment in a single step and ensure this keeps updating the original `df`.

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy

  df_applicant['NAME_EDUCATION_TYPE'][mask_EDU]='Other'
C:\Users\gibib\AppData\Local\Temp\ipykernel_14552\3122443786.py:3: Setting

In [12]:
#To change less occuring categories as 'other'
mask_HOU=df_applicant['NAME_HOUSING_TYPE'].isin(counts_NAME_HOUSING_TYPE[counts_NAME_HOUSING_TYPE<1000].index)
df_applicant['NAME_HOUSING_TYPE'][mask_HOU]='Other'
print(df_applicant['NAME_HOUSING_TYPE'].value_counts())

NAME_HOUSING_TYPE
House / apartment      32548
With parents            1776
Municipal apartment     1128
Other                   1005
Name: count, dtype: int64


C:\Users\gibib\AppData\Local\Temp\ipykernel_14552\2242230253.py:3: FutureWarning: ChainedAssignmentError: behaviour will change in pandas 3.0!
You are setting values through chained assignment. Currently this works in certain cases, but when using Copy-on-Write (which will become the default behaviour in pandas 3.0) this will never work to update the original DataFrame or Series, because the intermediate object on which we are setting values will behave as a copy.
A typical example is when you are setting values in a column of a DataFrame, like:

df["col"][row_indexer] = value

Use `df.loc[row_indexer, "col"] = values` instead, to perform the assignment in a single step and ensure this keeps updating the original `df`.

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy

  df_applicant['NAME_HOUSING_TYPE'][mask_HOU]='Other'
C:\Users\gibib\AppData\Local\Temp\ipykernel_14552\2242230253.py:3: SettingWi

Machine learning algorithms require their input data to be represented as a vector or a matrix and assume data should be normally distributed. So we have to represent all categories as numbers using Dummy Encoding.

In [13]:
#Creating dummies for category variables
df_applicant_dummies=pd.get_dummies(df_applicant, drop_first =True)

In [14]:
#Converting it to int datatype
df_applicant_dummies = df_applicant_dummies.astype(int)

In [15]:
df_applicant_dummies.head()

,ID,FLAG_OWN_CAR,FLAG_OWN_REALTY,CNT_CHILDREN,AMT_INCOME_TOTAL,DAYS_BIRTH,DAYS_EMPLOYED,FLAG_WORK_PHONE,FLAG_PHONE,FLAG_EMAIL,...,NAME_EDUCATION_TYPE_Incomplete higher,NAME_EDUCATION_TYPE_Other,NAME_EDUCATION_TYPE_Secondary / secondary special,NAME_FAMILY_STATUS_Married,NAME_FAMILY_STATUS_Separated,NAME_FAMILY_STATUS_Single / not married,NAME_FAMILY_STATUS_Widow,NAME_HOUSING_TYPE_Municipal apartment,NAME_HOUSING_TYPE_Other,NAME_HOUSING_TYPE_With parents
0,5008804,1,1,0,427500,-12005,-4542,1,0,0,...,0,0,0,0,0,0,0,0,1,0
1,5008805,1,1,0,427500,-12005,-4542,1,0,0,...,0,0,0,0,0,0,0,0,1,0
2,5008806,1,1,0,112500,-21474,-1134,0,0,0,...,0,0,1,1,0,0,0,0,0,0
3,5008808,0,1,0,270000,-19110,-3051,0,1,1,...,0,0,1,0,0,1,0,0,0,0
4,5008809,0,1,0,270000,-19110,-3051,0,1,1,...,0,0,1,0,0,1,0,0,0,0


In [16]:
#Dropping columns DAYS_BIRTH and DAYS_EMPLOYED to avoid negative no.s by creating two new columns.
df_applicant_dummies['AGE_YEARS'] = (-df_applicant_dummies['DAYS_BIRTH'] / 365).astype(int)
df_applicant_dummies['EMPLOYED_YEARS'] = (-df_applicant_dummies['DAYS_EMPLOYED'] / 365)
df_applicant_dummies.drop(columns=['DAYS_BIRTH','DAYS_EMPLOYED'], inplace=True)

In [17]:
df_applicant_dummies.head()

,ID,FLAG_OWN_CAR,FLAG_OWN_REALTY,CNT_CHILDREN,AMT_INCOME_TOTAL,FLAG_WORK_PHONE,FLAG_PHONE,FLAG_EMAIL,CNT_FAM_MEMBERS,STATUS,...,NAME_EDUCATION_TYPE_Secondary / secondary special,NAME_FAMILY_STATUS_Married,NAME_FAMILY_STATUS_Separated,NAME_FAMILY_STATUS_Single / not married,NAME_FAMILY_STATUS_Widow,NAME_HOUSING_TYPE_Municipal apartment,NAME_HOUSING_TYPE_Other,NAME_HOUSING_TYPE_With parents,AGE_YEARS,EMPLOYED_YEARS
0,5008804,1,1,0,427500,1,0,0,2,0,...,0,0,0,0,0,0,1,0,32,12.443836
1,5008805,1,1,0,427500,1,0,0,2,0,...,0,0,0,0,0,0,1,0,32,12.443836
2,5008806,1,1,0,112500,0,0,0,2,1,...,1,1,0,0,0,0,0,0,58,3.106849
3,5008808,0,1,0,270000,0,1,1,1,1,...,1,0,0,1,0,0,0,0,52,8.358904
4,5008809,0,1,0,270000,0,1,1,1,1,...,1,0,0,1,0,0,0,0,52,8.358904


In [18]:
df_applicant_dummies.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 36457 entries, 0 to 36456
Data columns (total 27 columns):
 #   Column                                             Non-Null Count  Dtype  
---  ------                                             --------------  -----  
 0   ID                                                 36457 non-null  int64  
 1   FLAG_OWN_CAR                                       36457 non-null  int64  
 2   FLAG_OWN_REALTY                                    36457 non-null  int64  
 3   CNT_CHILDREN                                       36457 non-null  int64  
 4   AMT_INCOME_TOTAL                                   36457 non-null  int64  
 5   FLAG_WORK_PHONE                                    36457 non-null  int64  
 6   FLAG_PHONE                                         36457 non-null  int64  
 7   FLAG_EMAIL                                         36457 non-null  int64  
 8   CNT_FAM_MEMBERS                                    36457 non-null  int64  
 9   STATUS

In [19]:
df_applicant_dummies.describe()

,ID,FLAG_OWN_CAR,FLAG_OWN_REALTY,CNT_CHILDREN,AMT_INCOME_TOTAL,FLAG_WORK_PHONE,FLAG_PHONE,FLAG_EMAIL,CNT_FAM_MEMBERS,STATUS,...,NAME_EDUCATION_TYPE_Secondary / secondary special,NAME_FAMILY_STATUS_Married,NAME_FAMILY_STATUS_Separated,NAME_FAMILY_STATUS_Single / not married,NAME_FAMILY_STATUS_Widow,NAME_HOUSING_TYPE_Municipal apartment,NAME_HOUSING_TYPE_Other,NAME_HOUSING_TYPE_With parents,AGE_YEARS,EMPLOYED_YEARS
count,3.645700e+04,36457.000000,36457.000000,36457.000000,3.645700e+04,36457.000000,36457.000000,36457.000000,36457.000000,36457.000000,...,36457.000000,36457.000000,36457.000000,36457.000000,36457.000000,36457.000000,36457.000000,36457.000000,36457.000000,36457.000000
mean,5.078227e+06,0.379708,0.672189,0.430315,1.866857e+05,0.225526,0.294813,0.089722,2.198453,0.882300,...,0.679623,0.687056,0.057684,0.132457,0.042022,0.030941,0.027567,0.048715,43.260334,7.247723
std,4.187524e+04,0.485321,0.469422,0.742367,1.017892e+05,0.417934,0.455965,0.285787,0.911686,0.322257,...,0.466628,0.463698,0.233149,0.338992,0.200642,0.173159,0.163730,0.215274,11.510414,5.890187
min,5.008804e+06,0.000000,0.000000,0.000000,2.700000e+04,0.000000,0.000000,0.000000,1.000000,0.000000,...,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,20.000000,0.046575
25%,5.042028e+06,0.000000,0.000000,0.000000,1.215000e+05,0.000000,0.000000,0.000000,2.000000,1.000000,...,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,34.000000,3.202740
50%,5.074614e+06,0.000000,1.000000,0.000000,1.575000e+05,0.000000,0.000000,0.000000,2.000000,1.000000,...,1.000000,1.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,42.000000,6.898630
75%,5.115396e+06,1.000000,1.000000,1.000000,2.250000e+05,0.000000,1.000000,0.000000,3.000000,1.000000,...,1.000000,1.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,53.000000,8.638356
max,5.150487e+06,1.000000,1.000000,19.000000,1.575000e+06,1.000000,1.000000,1.000000,20.000000,1.000000,...,1.000000,1.000000,1.000000,1.000000,1.000000,1.000000,1.000000,1.000000,68.000000,43.049315


In [20]:
#Checking for missing values
print(df_applicant_dummies.isna().sum())

ID                                                   0
FLAG_OWN_CAR                                         0
FLAG_OWN_REALTY                                      0
CNT_CHILDREN                                         0
AMT_INCOME_TOTAL                                     0
FLAG_WORK_PHONE                                      0
FLAG_PHONE                                           0
FLAG_EMAIL                                           0
CNT_FAM_MEMBERS                                      0
STATUS                                               0
CODE_GENDER_M                                        0
NAME_INCOME_TYPE_Pensioner                           0
NAME_INCOME_TYPE_State servant                       0
NAME_INCOME_TYPE_Student                             0
NAME_INCOME_TYPE_Working                             0
NAME_EDUCATION_TYPE_Incomplete higher                0
NAME_EDUCATION_TYPE_Other                            0
NAME_EDUCATION_TYPE_Secondary / secondary special    0
NAME_FAMIL

There are no missing values in the dataset.

In [21]:
# Drop identifier
df = df_applicant_dummies.drop(columns=['ID'])

In [22]:
# Remove duplicate applicants
df = df.drop_duplicates()

In [23]:
#Changing the order of STATUS column 
cols = ['STATUS'] + [c for c in df.columns if c != 'STATUS']
df = df[cols]

In [24]:
df.head()

,STATUS,FLAG_OWN_CAR,FLAG_OWN_REALTY,CNT_CHILDREN,AMT_INCOME_TOTAL,FLAG_WORK_PHONE,FLAG_PHONE,FLAG_EMAIL,CNT_FAM_MEMBERS,CODE_GENDER_M,...,NAME_EDUCATION_TYPE_Secondary / secondary special,NAME_FAMILY_STATUS_Married,NAME_FAMILY_STATUS_Separated,NAME_FAMILY_STATUS_Single / not married,NAME_FAMILY_STATUS_Widow,NAME_HOUSING_TYPE_Municipal apartment,NAME_HOUSING_TYPE_Other,NAME_HOUSING_TYPE_With parents,AGE_YEARS,EMPLOYED_YEARS
0,0,1,1,0,427500,1,0,0,2,1,...,0,0,0,0,0,0,1,0,32,12.443836
2,1,1,1,0,112500,0,0,0,2,1,...,1,1,0,0,0,0,0,0,58,3.106849
3,1,0,1,0,270000,0,1,1,1,0,...,1,0,0,1,0,0,0,0,52,8.358904
7,1,0,1,0,283500,0,0,0,1,0,...,0,0,1,0,0,0,0,0,61,7.246575
10,1,1,1,0,270000,1,1,1,2,1,...,0,1,0,0,0,0,0,0,46,2.106849


In [25]:
#Checking class imbalance
df['STATUS'].value_counts(normalize=True)

STATUS
1    0.786796
0    0.213204
Name: proportion, dtype: float64

When checked class imbalance, it was found that the dataset is imbalanced. 0.786796 proportion of customers are good customers and others are bad customers.

In [26]:
#Specifying X and y for splitting
X= df.drop("STATUS",axis=1).values
y= df["STATUS"].values

In [27]:
#Train test splitting
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state= 42, test_size=0.2, stratify =y)

In [28]:
#Standardizing
scaler = StandardScaler()
X_train_scaled=scaler.fit_transform(X_train)
X_test_scaled=scaler.transform(X_test)

All features were standardized for consistency. Binary and dummy variables were retained in standardized form as the subsequent models are scale-invariant.

In [29]:
# save the data to a new csv file
import numpy as np
df.to_csv(
    r'C:\Users\gibib\Desktop\Data Scientist\Springboard-Project-Capstone-2\Credit Card Data\Applicant_data_trained.csv',
    index=False
)
np.save(r'C:\Users\gibib\Desktop\Data Scientist\Springboard-Project-Capstone-2\Credit Card Data\X_train_scaled.npy', X_train_scaled)
np.save(r'C:\Users\gibib\Desktop\Data Scientist\Springboard-Project-Capstone-2\Credit Card Data\X_test_scaled.npy', X_test_scaled)
np.save(r'C:\Users\gibib\Desktop\Data Scientist\Springboard-Project-Capstone-2\Credit Card Data\y_train.npy', y_train)
np.save(r'C:\Users\gibib\Desktop\Data Scientist\Springboard-Project-Capstone-2\Credit Card Data\y_test.npy', y_test)

SUMMARY:
During Preprocessing_and_training, at first there were 17 columns. The dataset was cleaned, transformed, encoded(resulting in adding 10 more columns), and split into training and testing subsets. Feature scaling was applied following the split to prevent data leakage. The resulting datasets are suitable for the modeling phase.